### PR TITLE
fix(ci): improve test-coverage-check.yml and check_test_coverage.py — 3 edge case fixes

### PR DESCRIPTION
**Pull request by anyang-keepers**
## Summary
Implements 3 of the 4 edge cases from qa-king's review (board p-6e011a6536b4155280c2fb71d40b9269) for the CI test coverage gate. Edge case 4 (repo-level path rubric) is out of scope for the script — deferred to v2.

### Edge case 1 — expanded trigger paths
`test-coverage-check.yml` now triggers on `dashboard/**` and `config/**` changes in addition to `lib/**`. The coverage check was previously silent for non-lib changes.

### Edge case 2 — scan-all-commits opt-out
`is_opt_out_commit()` now scans all PR commits (`origin/BASE_REF...HEAD`) instead of only the latest commit (`git log -1`). Falls back to latest commit when outside CI.

### Edge case 3 — broader test file glob
`get_changed_test_files()` adds `*check_*`, `*validator*`, and `**/test/` pathspecs to catch `check_*.ml`, `*validator*.ml`, and `test/` directory patterns.

### v2 deferred
Edge case 4 (repo-level path rubric in function-level granularity guard) requires a different approach — deferred to future work.

Closes #20921

ref p-6e011a6536b4155280c2fb71d40b9269

### CODE REVIEW

**Comment by jeong-sik** [comment]
### Review Response

@chatgpt-codex-connector의 3건 P2 리뷰를 확인했습니다:

1. **P2 opt-out PR-side 제한**: base branch commit까지 스캔하면 false positive. PR diff 기준으로 제한하는 것이 맞습니다.
2. **P2 validator/check glob 제한**: 프로덕션 코드도 매칭되는 broad glob. test path로 좁히는 것이 안전합니다.
3. **P2 dashboard/config coverage 확장**: dashboard 경로가 trigger에 누락되어 있어 coverage check가 돌지 않습니다.

3건 모두 개선 사항으로 보입니다. 다만 이 PR의 CI가 이미 pass 중이므로, follow-up PR로 반영하는 것이 안전합니다.

File: scripts/check_test_coverage.py [hunk 1]
```diff
@@ -19,17 +19,35 @@
...
+    check for earlier commits (edge case: opt-out buried in the last
+    commit while substantive changes sit in earlier commits).
+    Falls back to the latest commit alone when the base ref is
+    unavailable (local runs outside CI).
+    """
+    base_ref = os.environ.get("GITHUB_BASE_REF", "main")
     try:
         result = subprocess.run(
-            ["git", "log", "-1", "--format=%s%n%b"],
+            ["git", "log", f"origin/{base_ref}...HEAD", "--format=%s%n%b"],
```
[the remark this answers is not in the document]
Comment (jeong-sik) [reply scripts/check_test_coverage.py]: Fixed in 070d5c21a: commit-message opt-out scanning now uses the one-sided PR commit range (`origin/$GITHUB_BASE_REF..HEAD`) so base-only commits cannot opt out stale PRs. Added a unit test for the exact git log range.

File: scripts/check_test_coverage.py [hunk 2]
```diff
@@ -70,6 +93,9 @@ def get_changed_test_files():
             "--",
             "*test*",
             "*spec*",
+            "*check_*",
+            "*validator*",
```
[the remark this answers is not in the document]
Comment (jeong-sik) [reply scripts/check_test_coverage.py]: Fixed in 070d5c21a: replaced the broad `*check_*` / `*validator*` pathspecs with an `is_test_file()` predicate. Production check/validator modules no longer self-satisfy coverage; unit tests cover the cited production paths and valid test paths.

File: .github/workflows/test-coverage-check.yml [hunk 3]
```diff
@@ -3,6 +3,8 @@ on:
   pull_request:
     paths:
       - 'lib/**'
+      - 'dashboard/**'
+      - 'config/**'
```
[the remark this answers is not in the document]
Comment (jeong-sik) [reply .github/workflows/test-coverage-check.yml:7]: Fixed in 070d5c21a: the checker now treats `lib/`, `dashboard/`, and `config/` as covered code paths for changed-file and added-line violations, matching the workflow triggers. Added a unit test that dashboard-only changes without tests fail the gate.

**Review by jeong-sik — APPROVED** [review]
Approved after addressing the three coverage-gate review gaps in 070d5c21a. Verified locally with py_compile, unittest, ruff, pyright, and git diff --check; review threads are resolved. Merge after the remaining GitHub checks finish green.
